### ISSUE TITLE
Improved RAM offloading algorithm

### ISSUE THREAD

**Issue by Arcitec** [kohya-ss/musubi-tuner#581]
Hi, in case it helps, OneTrainer invented a very fast RAM offloading technique and has documented it here:

https://github.com/Nerogar/OneTrainer/blob/master/docs/RamOffloading.md

The general strategy is:

- Allocate a large, contiguous block of VRAM. This stays permanently allocated, to avoid many years old, still unfixed PyTorch VRAM allocation bugs and overhead. Otherwise you get fragmentation and can get false "OOMs".
- The large block of VRAM is then sliced manually by OneTrainer by uploading model layers into it and casting those pointers to be the correct datatypes on-demand.
- The uploading of data is done asynchronously via "pinned" system RAM, which gives the NVIDIA driver much faster access to transfer system RAM data into VRAM. The GPU itself reads that RAM via DMA, if I remember right.
- The upload sequencing is done via an algorithm that always asynchronously uploads one or more of the "next" layers. It's uploaded in parallel while the GPU is still working on previous model layers, so that the GPU doesn't have to wait for the uploads. When the GPU is ready to use the next layer, it's already in VRAM.

As a result, it performs RAM offloading with almost no performance loss compared to having a lot of real VRAM.

**Comment by kohya-ss** [comment]
Thank you for this! I know OneTrainer has a great offloading mechanism, and I think it offers the best performance. However, it seems a bit too complicated for me and I feel it would be difficult to maintain properly even if I were to merge the functionality into this repository.

Another offloading mechanism with DMA transfers, asynchronous transfers, and reduced memory allocations, similar to OneTrainer, is implemented in this repository, based on an idea by 2kpr, and called "block swap".

The code is very simple, but I think it's about 90% as fast as OneTrainer, so I feel it's sufficient.

I'm also working on reducing the shared GPU memory usage, which, once implemented, should further reduce memory requirements.

**Comment by Arcitec** [comment]
@kohya-ss Thank you so much for the explanation. After hearing what you're already doing here, I agree with you. Your method is doing most of the important things (async/DMA/reduced allocations), and it sounds like it will be almost as fast but more compatible, without needing to struggle as much about how to implement layer strategies for every individual model (which I know was hard for OneTrainer).

I am so incredibly impressed with your project, by the way. You've clearly taken years of experience from sd-scripts and done absolutely everything with beautifully modern code.

- uv for robust package management, with pinned requirements to avoid bugs with people getting random dependency versions.
- Beautiful, modern `src/` module organization, with great sub-module structure.
- The code is nicely split into small, nicely commented, well-named functions.
- Clean code formatting and linting with ruff.
- You have a fantastic RAM offloading algorithm, which is a huge help as long as NVIDIA won't give us 96 GB consumer GPUs. ;)
- It looks like there might even be support for DeepSpeed which can be a benefit for training some models.

Even though your project is young, it has quickly picked up a lot of repeat contributors. Definitely thanks to your reputation and the clean codebase. What an awesome project! I really admire what you have created. This is turning out to be the ultimate AI training toolkit. The one and only to rule them all. 😸 

![Image](https://github.com/user-attachments/assets/fc2326ad-2b84-419a-b42f-bbe365671bbf)

**Comment by kohya-ss** [comment]
Yes, the offloader seems to be working well so far.

If this repository is well organized, it's thanks to the contributors and the lessons learned from sd-scripts. uv and ruff in particular wouldn't have been implemented without the cooperation of the contributors. I'd like to express my gratitude to them again.

Thank you for choosing Musubi Tuner!

Each trainer has its own characteristics, so I think it's best for users to use the one that's right for them (for example, I recommend diffusion-pipe for full finetuning).

But even so, I'd be happy if you still use Musubi Tuner😀

**Comment by Sarania** [comment]
Welcome! I gotta say personally I use this offloader on somewhat older hardware (x99 board = DDR4 though it is quad channel at least, also PCIExpress 3 for my poor 4070 TS but at least it gets all 16 lanes I've got 40 PCIe lanes so even with two GPUs and an NVME SSD they all still get full bandwidth!) and recently calculated that my overhead to swap the entirety of a 33GB model one time is about 4.2 seconds with a real world workload! I'm certain with newer hardware e.g. DDR5 + PCIe 5 it would do a fair chunk better than that as well. Also I gotta say another thing I like about it is it's super easy to incorporate into new models I've done so with several myself. Then add in easy support for both inference and training style tasks... it's my favorite bit of code in Musubi after fp8_scaled!

**Comment by Arcitec** [comment]
@kohya-ss @Sarania Thank you both and everyone else that has been involved in creating Musubi Tuner. I love it. It's so well-structured and so intelligent. Since I don't need full fine-tuning, it will be the only tool I ever need! 😸 💯 

It's so cool to hear from Sarania that it was easy to add more models. This project has such great structure!
